Germline variant caller that does a local reassembly

arahuja opened the pull request:
The basic summary of this caller is:
- Find areas where > some % of the reads show a variant
- Rexamine those areas with N base window around it
- Place the reads in the N-base window into a DeBruijn graph and find paths between the start and end of the reference sequence that covers that region
- Align those paths to the reference sequence to discover variants

This code update `DeBruijnGraph` code to backtrack on dead-end paths as well as some helper functions to `ReadAlignment`

Right now there is a single test based on the Illumina NA12878 platinum extract that @timodonnell created.

There are more updates to come (in future PRs) but this is the first full working version.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/hammerlab/guacamole/372)
<!-- Reviewable:end -->

timodonnell commented:
Very cool. I didn't look at this in detail but my only suggestion is to add the basic summary you gave in this PR as a comment somewhere (maybe in GermlineAssemblyCaller.scala). 

Unit tests for the caller on specific cases would be good to have in the final version, but I like your idea to merge in this caller piecemeal so we avoid diverging even more on our separate branches. LGTM

ryan-williams commented on src/main/scala/org/hammerlab/guacamole/assembly/DeBruijnGraph.scala:
```diff
@@ -213,13 +215,20 @@ class DeBruijnGraph(val kmerSize: Int,
       } else {
-        //found sink or too long
+        // Found sink or too long
         if (foundSink && currentPath.size + 1 >= minPathLength) {
```
I'm confused by the `currentPath.size + 1 >= minPathLength` check here; don't you want to skip if the path length is greater than `maxPathLength`?

arahuja replied:
Maybe the nesting is confusing here, but what should happen is, we should end up here if we either found our goal or we cut-off the search. This check that if we did find the sink/goal, the path was long enough.  Will clarify some comments, but let me know if that seems wrong

ryan-williams commented on src/main/scala/org/hammerlab/guacamole/assembly/DeBruijnGraph.scala:
```diff
@@ -213,13 +215,20 @@ class DeBruijnGraph(val kmerSize: Int,
+
+        // Track if this is a branching node
+        if (filteredNextNodes.size > 1) {
+          lastBranchIndex = currentPath.length
```
can there be nested branches? should this keep a stack of branch indices?

arahuja replied:
You are right! A stack would fix this, along with checking how often we back-tracked.

ryan-williams commented on src/main/scala/org/hammerlab/guacamole/assembly/DeBruijnGraph.scala:
```diff
@@ -293,9 +304,9 @@ object DeBruijnGraph {
-    val rest = kmers.map(_.last)
+  def mergeKmers(kmers: Seq[Kmer], kmerSize: Int): Sequence = {
+    val head = kmers.headOption.getOrElse(Seq.empty)
+    val rest = kmers.tail.flatMap(kmer => kmer.takeRight(kmer.length - kmerSize + 1))
```
so I guess you've adapted this to work on merged "k-mers"?

Calling merged-kmers (that are longer than k) "kmers" makes this a little confusing; can we try to distinguish the var names or types, or comment it at all?

arahuja replied:
Yea, I think that makes sense. I think I will change this to 'mergeOverlaping` or something of that sort and specify an overlap size?

ryan-williams replied:
sure I think that would be clearer, maybe a comment about it too?

ryan-williams commented on src/main/scala/org/hammerlab/guacamole/commands/GermlineAssemblyCaller.scala:
```diff
@@ -0,0 +1,271 @@
+      val referenceKmerSink = currentReference.takeRight(kmerSize)
+
+      // If the current window size doesn't cover the kmer size we won't be able to find the reference start/end
+      if (referenceKmerSource.length != kmerSize || referenceKmerSink.length != kmerSize)
```
is this the same as saying `currentReference.size < kmerSize`?

arahuja replied:
Yea, that also works, seems more concise will change

ryan-williams commented on src/main/scala/org/hammerlab/guacamole/commands/GermlineAssemblyCaller.scala:
```diff
@@ -0,0 +1,271 @@
+
+      // Take `ploidy` paths
+      val topPaths = paths
+        .take(expectedPloidy)
```
won't this take the two longest paths found? is that intentional?

arahuja replied:
Yea, not even necessarily the two longest but just any two really. Need to come up with some sort of weighting/scoring strategy. I think the better solution will be to take some larger number (6-8) and align them and takes the ones with the best alignments.

ryan-williams replied:
right, DFS, so random two paths.

aligning all and choosing best two seems like the correct output.. but it seems like we could easily end up with exponentially many paths easily; hard-stopping at `maxNumPaths` isn't a great solution. I wonder what the right way to do this is.

ryan-williams commented on src/main/scala/org/hammerlab/guacamole/commands/GermlineAssemblyCaller.scala:
```diff
@@ -0,0 +1,271 @@
+        val mappingQualities = DenseVector(reads.map(_.alignmentQuality.toFloat).toArray)
+        val baseQualities =  DenseVector(reads.flatMap(_.baseQualities).map(_.toFloat).toArray)
+        CalledAllele(
+          reads.head.sampleName,
```
minor nit but maybe store these in `sampleName` and `referenceContig` `val`s outside of this method / the loops that call it?

ryan-williams commented on src/main/scala/org/hammerlab/guacamole/commands/GermlineAssemblyCaller.scala:
```diff
@@ -0,0 +1,271 @@
+     * @param expectedPloidy Expected ploidy, or expected number of valid paths through the graph
+     * @return A set of variants in the window
+     */
+    def discoverHaplotypes(graph: Option[DeBruijnGraph],
```
`graph` is never modified in any way and is simply returned, afaict; is there a reason to include it as a param and return value?

arahuja replied:
This was part of a plan to have the graph updated each time rather than rebuilt each time, but this was not implemented. I can simplify this more now and remove it until it is actually implemented.

ryan-williams replied:
gotcha, sorry if I missed that somewhere. I'm fine leaving it in but maybe a TODO comment somewhere seems warranted

ryan-williams commented:
This is awesome @arahuja. I've left some comments/questions, lmk what you think.

timodonnell commented on src/main/scala/org/hammerlab/guacamole/commands/GermlineAssemblyCaller.scala:
```diff
@@ -0,0 +1,271 @@
+    /**
+     * Extract alignments that are mismatches, insertions or deletions
+     * @param alignment
+     * @return Filtered set of alignments that are mismatches, insertions or deletions
```
nit: it seems to be returning a Seq not Set, and also pairs, might be worth clarifying what the second element of the pair is

timodonnell commented on src/main/scala/org/hammerlab/guacamole/commands/GermlineAssemblyCaller.scala:
```diff
@@ -0,0 +1,271 @@
+    }
+
+    /**
+     *
```
An overview sentence or two on what this does would be helpful

arahuja commented:
Thanks @timodonnell and @ryan-williams for all the feedback! Updated with comments where necessary as well as some of the various simplifications. @ryan-williams thanks for pointing out the issue with the multiple splits in the graph that was not handled. I added the stack as suggested and a test case where we need to backtrack twice to find the right path.

ryan-williams commented on src/main/scala/org/hammerlab/guacamole/assembly/DeBruijnGraph.scala:
```diff
@@ -213,14 +219,21 @@ class DeBruijnGraph(val kmerSize: Int,
-        //found sink or too long
-        if (foundSink && currentPath.size + 1 >= minPathLength) {
+      } else { // Found sink or search was cut short
+        if (foundSink && currentPath.size + 1 >= minPathLength) { // Found sink with valid length
```
thanks for the added comment; forgot to ask before: why `+ 1` here?

ryan-williams commented:
Thanks @arahuja; the big open question seems to be about scoring paths vs. returning two random ones.. do you want to tackle that here or later? I'm fine either way.

There are a few tests I could see adding also, e.g. a case where a valid path is found as the third path from a branch-point. Up to you though.

arahuja commented:
> the big open question seems to be about scoring paths vs. returning two random ones.. do you want to tackle that here or later? I'm fine either way.

Yea for this, I'd like to get some more experience with this on real data - basically how many valid paths are usually found. (My suspusicion is it will be small unless we are in a repetitive region, but then we should find a way to identify to those and do something else)

>a case where a valid path is found as the third path from a branch-point. Up to you though.
Not sure I get this test case exactly, but I think we can add these as we find some interesting cases. If you have some specific examples that might fail, please add them or let me know though.

ryan-williams commented:
> basically how many valid paths are usually found. (My suspusicion is it will be small…

It seems to me that any single-bp seq-error at the location of a het SNV will give a 2/3 chance of not picking the right 2 (of 3) paths, with the current approach; is that consistent with your understanding?

arahuja commented:
> It seems to me that any single-bp seq-error at the location of a het SNV will give a 2/3 chance of not picking the right 2 (of 3) paths, with the current approach; is that consistent with your understanding?

Yes, but `minOccurence` should prune out most of the simpler cases.

ryan-williams commented:
Fair enough; do we have any way of learning "how many valid paths are usually found"? Should we log a warning or something if it's >2? Or how will we know?

arahuja commented:
> Should we log a warning or something if it's >2? Or how will we know?

I was thinking we would just find some interesting variants and start testing on them to get a sense for performance. With that we can develop a reasonable scoring strategy.

ryan-williams commented:
I think we're agreed on these two things:

* if there are >2 paths, this code will do the wrong thing with >50% probability.
* in a germline-calling scenario this should ~never happen.

Given this, I think we should throw an exception. Occasionally outputting variants we know are most likely wrong, and having to debug this later starting from such calls, sounds like a bad time.

arahuja commented:
>  I think we should throw an exception.  
I don't want to throw an exception as this would halt execution on a dataset where it did happen and make it difficult to examine the rest of the results.

I've added a parameter to score them if there are less them `maxNumberPathsToScore` and to otherwise output no variants and log a warning.

ryan-williams commented:
thanks, lgtm
